pkg/operator/controller/ingress/deployment.go

```diff
@@ -75,40 +78,102 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
+		// at the completion of the upgrade.  To this end, we add a
+		// label with a hash of the deployment, using which we can
+		// select replicas of the same generation (or select replicas
+		// that are *not* of the same generation), configure affinity to
+		// colocate replicas of different generations of the same
+		// ingress controller, and configure anti-affinity to prevent
+		// colocation of replicas of the same generation of the same
+		// ingress controller.
+		needDeploymentHash = true
+		deployment.Spec.Template.Spec.Affinity = &corev1.Affinity{
+			PodAffinity: &corev1.PodAffinity{
+				PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{
+					{
+						Weight: int32(100),
+						PodAffinityTerm: corev1.PodAffinityTerm{
+							TopologyKey: "kubernetes.io/hostname",
+							LabelSelector: &metav1.LabelSelector{
+								MatchExpressions: []metav1.LabelSelectorRequirement{
+									{
+										Key:      controller.ControllerDeploymentLabel,
+										Operator: metav1.LabelSelectorOpIn,
+										Values:   []string{controller.IngressControllerDeploymentLabel(ci)},
+									},
+									{
+										Key:      controller.ControllerDeploymentHashLabel,
+										Operator: metav1.LabelSelectorOpNotIn,
+										// Values is set at the end of this function.
+									},
+								},
 							},
 						},
 					},
 				},
 			},
-		},
-	}
+			PodAntiAffinity: &corev1.PodAntiAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+					{
+						TopologyKey: "kubernetes.io/hostname",
+						LabelSelector: &metav1.LabelSelector{
+							MatchExpressions: []metav1.LabelSelectorRequirement{
+								{
+									Key:      controller.ControllerDeploymentLabel,
+									Operator: metav1.LabelSelectorOpIn,
+									Values:   []string{controller.IngressControllerDeploymentLabel(ci)},
+								},
+								{
+									Key:      controller.ControllerDeploymentHashLabel,
+									Operator: metav1.LabelSelectorOpIn,
+									// Values is set at the end of this function.
+								},
+							},
+						},
+					},
+				},
+			},
+		}
 
-	// For now, all strategies use 25% max unavailable and 0 surge. This is because
-	// distinct ingress controllers can't currently be colocated. Usually, replicas
-	// will be equal to the node pool size. Under these conditions, surge requires
-	// new nodes to support the rollout. This means a positive surge can cause the
-	// rollout to wedge in the absence of auto-scaling.
-	pointerTo := func(ios intstr.IntOrString) *intstr.IntOrString { return &ios }
-	deployment.Spec.Strategy = appsv1.DeploymentStrategy{
-		Type: appsv1.RollingUpdateDeploymentStrategyType,
-		RollingUpdate: &appsv1.RollingUpdateDeployment{
-			MaxUnavailable: pointerTo(intstr.FromString("25%")),
-			MaxSurge:       pointerTo(intstr.FromInt(0)),
-		},
+		// Avoid going from one replica to zero replicas on any given node.
```

Comment:
Would you mind also talking about the details of deployments that explain how these values accomplish the stated behavior?

Reply:
I have updated the comments better to tell the tale of Ingress Operator and the Zero-Downtime Rolling Update.